modules/core/src/test/java/org/apache/ignite/internal/processors/cache/NotAffinitySupplierWithMultipalRebalanceTest.java

```diff
@@ -0,0 +1,403 @@
+    public void testPersistentFullRebalance() throws Exception {
+        supplingFromOldBackup(true);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testInMemoryFullRebalance() throws Exception {
+        supplingFromOldBackup(false);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testPersistentHistoricalRebalance() throws Exception {
+        supplingHistoricalFromOldBackup(true);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void supplingHistoricalFromOldBackup(boolean persistentEnabled) throws Exception {
+        System.setProperty(IGNITE_PDS_WAL_REBALANCE_THRESHOLD, "0");
+
+        this.persistentEnabled = persistentEnabled;
+
+        IgniteEx ignite0 = startGrids(NODES_CNT);
+
+        for (int i = 1; i < 4; i++)
+            startGrid("new_" + i);
+
+        ignite0.cluster().active(true);
+
+        loadData(ignite0, RENDEZVOUS_CACHE);
+        loadData(ignite0, CUSTOM_CACHE);
+
+        ignite0.cluster().baselineAutoAdjustEnabled(false);
+
+        awaitPartitionMapExchange();
+
+        stopGrid("new_1");
+        stopGrid("new_2");
+
+        awaitPartitionMapExchange();
+
+        loadData(ignite0, RENDEZVOUS_CACHE);
+        loadData(ignite0, CUSTOM_CACHE);
+
+        TestRecordingCommunicationSpi testCommunicationSpi1 = startNodeWithBlockingRebalance("new_1", true);
+        TestRecordingCommunicationSpi testCommunicationSpi2 = startNodeWithBlockingRebalance("new_2", true);
+
+        testCommunicationSpi1.waitForBlocked();
+        testCommunicationSpi2.waitForBlocked();
+
+        checkPartitionsState(GridDhtPartitionState.RENTING, RENDEZVOUS_CACHE);
+        checkPartitionsState(GridDhtPartitionState.RENTING, CUSTOM_CACHE);
+
+        AtomicBoolean hasMissed = new AtomicBoolean();
+
+        for (Ignite ign : G.allGrids()) {
+            ign.events().localListen(event -> {
+                info("Partition missing event: " + event);
+
+                hasMissed.compareAndSet(false, true);
+
+                return false;
+            }, EVT_CACHE_REBALANCE_PART_MISSED);
+        }
+
+        TestRecordingCommunicationSpi testCommunicationSpi0 = (TestRecordingCommunicationSpi)ignite0
+            .configuration().getCommunicationSpi();
+
+        testCommunicationSpi0.record(GridDhtPartitionsFullMessage.class);
+
+        testCommunicationSpi1.stopBlock();
+
+        testCommunicationSpi0.waitForRecorded();
+
+        checkPartitionsState(GridDhtPartitionState.RENTING, CUSTOM_CACHE);
+
+        testCommunicationSpi2.stopBlock();
+
+        awaitPartitionMapExchange();
+
+        checkPartitionsState(GridDhtPartitionState.MOVING, RENDEZVOUS_CACHE);
+        checkPartitionsState(GridDhtPartitionState.MOVING, CUSTOM_CACHE);
+
+        assertFalse(hasMissed.get());
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void supplingFromOldBackup(boolean persistentEnabled) throws Exception {
```

Comment:
persistence instead of persistent i hope

Reply:
ok